File: 0077-combinations/0077-combinations.java
class Solution {
    private void dfs(List<Integer> path, List<List<Integer>> answer, int n, int k, int start) {
        if(path.size() == k) {
            answer.add(new ArrayList<>(path));
            return;
        }
        
        for(int i = start; i < n; i++) {
            path.add(i + 1);
            dfs(path, answer, n, k, i + 1);
            path.remove(Integer.valueOf(i + 1));
        }
    }
    
    public List<List<Integer>> combine(int n, int k) {
        List<List<Integer>> answer = new ArrayList<>();
        
        dfs(new ArrayList<>(), answer, n, k, 0);
        return answer;
    }
}

File: 0077-combinations/0077-combinations.kt
class Solution {
    fun combine(n: Int, k: Int): List<List<Int>> {
        // 결과 저장 리스트 선언
        val results: MutableList<List<Int>> = mutableListOf()
        
        // 중첩함수로 선언
        fun dfs(elements: LinkedList<Int>, start: Int, k: Int) {
            // k개의 조합을 모두 찾은 경우
            if(k == 0) {
                results.add(elements.toList())
            }
            
	        // 현재 엘리먼트 이후 인덱스부터 이후 엘리먼트 탐색
            for(i in start..n) {
                elements.add(i)
                dfs(elements, i + 1, k - 1)
                elements.removeLast()
            }
        }
        
        // dfs 시작
        dfs(LinkedList(), 1, k)
        return results
    }
}